File: src/lexer/lexer.py
from __future__ import annotations

from src.errors import LexerError, SourceLocation, SourceSpan, register_source
from src.lexer.tokens import KEYWORDS, Token, TokenKind


class Lexer:
    """Tokenises a single Vela source file."""

    def __init__(self, source: str, filename: str = "<stdin>") -> None:
        self._src = source
        self._file = filename
        register_source(filename, source)
        self._pos = 0
        self._line = 1
        self._col = 1
        self._tokens: list[Token] = []

    def tokenize(self) -> list[Token]:
        """Return the full list of tokens (including final EOF)."""
        while not self._at_end():
            self._skip_whitespace_and_comments()
            if self._at_end():
                break
            self._scan_token()
        self._tokens.append(self._make(TokenKind.EOF, ""))
        return self._tokens

    def _loc(self) -> SourceLocation:
        return SourceLocation(self._file, self._line, self._col)

    def _at_end(self) -> bool:
        return self._pos >= len(self._src)

    def _peek(self, offset: int = 0) -> str:
        i = self._pos + offset
        return self._src[i] if i < len(self._src) else "\0"

    def _advance(self) -> str:
        ch = self._src[self._pos]
        self._pos += 1
        if ch == "\n":
            self._line += 1
            self._col = 1
        else:
            self._col += 1
        return ch

    def _match(self, expected: str) -> bool:
        if self._at_end() or self._src[self._pos] != expected:
            return False
        self._advance()
        return True

    def _make(self, kind: TokenKind, value: str, loc: SourceLocation | None = None) -> Token:
        loc = loc or self._loc()
        span = SourceSpan(
            file=loc.file,
            start_line=loc.line,
            start_column=loc.column,
            end_line=max(loc.line, self._line),
            end_column=max(1, self._col),
        ).normalized()
        return Token(kind, value, loc, span)

    def _skip_whitespace_and_comments(self) -> None:
        while not self._at_end():
            ch = self._peek()
            if ch in (" ", "\t", "\r", "\n"):
                self._advance()
            elif ch == "/" and self._peek(1) == "/":
                while not self._at_end() and self._peek() != "\n":
                    self._advance()
            elif ch == "#":
                while not self._at_end() and self._peek() != "\n":
                    self._advance()
            else:
                break

    def _scan_token(self) -> None:
        loc = self._loc()
        ch = self._peek()

        # two-char tokens first
        if ch == "[" and self._peek(1) == "[":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.TAG_OPEN, "[[", loc))
            return
        if ch == "]" and self._peek(1) == "]":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.TAG_CLOSE, "]]", loc))
            return
        if ch == "=" and self._peek(1) == "=":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.EQ, "==", loc))
            return
        if ch == "!" and self._peek(1) == "=":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.NEQ, "!=", loc))
            return
        if ch == "<" and self._peek(1) == "=":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.LTE, "<=", loc))
            return
        if ch == ">" and self._peek(1) == "=":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.GTE, ">=", loc))
            return
        if ch == "&" and self._peek(1) == "&":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.AND, "&&", loc))
            return
        if ch == "|" and self._peek(1) == "|":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.OR, "||", loc))
            return
        if ch == "+" and self._peek(1) == "+":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.PLUS_PLUS, "++", loc))
            return
        if ch == "-" and self._peek(1) == "-":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.MINUS_MINUS, "--", loc))
            return
        if ch == "+" and self._peek(1) == "=":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.PLUS_EQ, "+=", loc))
            return
        if ch == "-" and self._peek(1) == "=":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.MINUS_EQ, "-=", loc))
            return
        if ch == "*" and self._peek(1) == "=":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.STAR_EQ, "*=", loc))
            return
        if ch == "/" and self._peek(1) == "=":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.SLASH_EQ, "/=", loc))
            return
        if ch == "<" and self._peek(1) == "-":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.ARROW, "<-", loc))
            return
        if ch == ":" and self._peek(1) == ":":
            self._advance(); self._advance()
            self._tokens.append(self._make(TokenKind.DOUBLE_COLON, "::", loc))
            return

        # single-char tokens
        single: dict[str, TokenKind] = {
            "+": TokenKind.PLUS, "-": TokenKind.MINUS,
            "*": TokenKind.STAR, "/": TokenKind.SLASH,
            "%": TokenKind.PERCENT, "!": TokenKind.NOT,
            "&": TokenKind.AMPERSAND, "=": TokenKind.ASSIGN,
            "<": TokenKind.LT, ">": TokenKind.GT,
            "(": TokenKind.LPAREN, ")": TokenKind.RPAREN,
            "{": TokenKind.LBRACE, "}": TokenKind.RBRACE,
            "[": TokenKind.LBRACKET, "]": TokenKind.RBRACKET,
            ";": TokenKind.SEMICOLON, ",": TokenKind.COMMA,
            ":": TokenKind.COLON, ".": TokenKind.DOT,
        }
        if ch in single:
            self._advance()
            self._tokens.append(self._make(single[ch], ch, loc))
            return

        # String literal
        if ch == '"':
            self._scan_string(loc)
            return

        # Char literal
        if ch == "'":
            self._scan_char(loc)
            return

        # Number literal (int or float)
        if ch.isdigit():
            self._scan_number(loc)
            return

        # Hex literal 0x...
        if ch == "0" and self._peek(1) in ("x", "X"):
            self._scan_number(loc)
            return

        # Identifier or keyword
        if ch.isalpha() or ch == "_":
            self._scan_identifier(loc)
            return

        raise LexerError(f"unexpected character {ch!r}", loc, span=SourceSpan.from_location(loc))

    def _scan_string(self, loc: SourceLocation) -> None:
        self._advance()  # opening "
        buf: list[str] = []
        while not self._at_end() and self._peek() != '"':
            if self._peek() == "\\":
                self._advance()
                esc = self._advance()
                buf.append({"n": "\n", "t": "\t", "\\": "\\", '"': '"', "0": "\0"}.get(esc, esc))
            else:
                buf.append(self._advance())
        if self._at_end():
            raise LexerError(
                "unterminated string literal",
                loc,
                span=SourceSpan.from_location(loc),
                hint="close the string literal with a double quote",
            )
        self._advance()  # closing "
        self._tokens.append(self._make(TokenKind.STRING_LITERAL, "".join(buf), loc))

    def _scan_char(self, loc: SourceLocation) -> None:
        self._advance()  # opening '
        if self._peek() == "\\":
            self._advance()
            ch = self._advance()
            val = {"n": "\n", "t": "\t", "\\": "\\", "'": "'", "0": "\0"}.get(ch, ch)
        else:
            val = self._advance()
        if self._at_end() or self._peek() != "'":
            raise LexerError(
                "unterminated char literal",
                loc,
                span=SourceSpan.from_location(loc),
                hint="close the character literal with a single quote",
            )
        self._advance()  # closing '
        self._tokens.append(self._make(TokenKind.CHAR_LITERAL, val, loc))

    def _scan_number(self, loc: SourceLocation) -> None:
        start = self._pos
        # Hex
        if self._peek() == "0" and self._peek(1) in ("x", "X"):
            self._advance(); self._advance()
            while not self._at_end() and self._peek() in "0123456789abcdefABCDEF_":
                self._advance()
            self._tokens.append(self._make(TokenKind.INT_LITERAL, self._src[start:self._pos], loc))
            return
        # Binary 0b
        if self._peek() == "0" and self._peek(1) in ("b", "B"):
            self._advance(); self._advance()
            while not self._at_end() and self._peek() in "01_":
                self._advance()
            self._tokens.append(self._make(TokenKind.INT_LITERAL, self._src[start:self._pos], loc))
            return
        # Decimal / float
        while not self._at_end() and (self._peek().isdigit() or self._peek() == "_"):
            self._advance()
        is_float = False
        if not self._at_end() and self._peek() == "." and self._peek(1) != ".":
            is_float = True
            self._advance()
            while not self._at_end() and (self._peek().isdigit() or self._peek() == "_"):
                self._advance()
        # Exponent
        if not self._at_end() and self._peek() in ("e", "E"):
            is_float = True
            self._advance()
            if not self._at_end() and self._peek() in ("+", "-"):
                self._advance()
            while not self._at_end() and self._peek().isdigit():
                self._advance()
        text = self._src[start:self._pos]
        kind = TokenKind.FLOAT_LITERAL if is_float else TokenKind.INT_LITERAL
        self._tokens.append(self._make(kind, text, loc))

    def _scan_identifier(self, loc: SourceLocation) -> None:
        start = self._pos
        while not self._at_end() and (self._peek().isalnum() or self._peek() == "_"):
            self._advance()
        text = self._src[start:self._pos]
        kind = KEYWORDS.get(text, TokenKind.IDENTIFIER)
        self._tokens.append(self._make(kind, text, loc))

File: ext/vela-lsp/src/vela/lexer.ts
import { DiagnosticSeverity } from "vscode-languageserver/node";
import {
  KEYWORDS,
  Token,
  TokenKind,
  VelaDiagnostic,
  VelaPosition,
  makePosition,
  makeRange,
} from "./model.js";

const TWO_CHAR_TOKENS: Record<string, TokenKind> = {
  "[[": TokenKind.TagOpen,
  "]]": TokenKind.TagClose,
  "==": TokenKind.Eq,
  "!=": TokenKind.Neq,
  "<=": TokenKind.Lte,
  ">=": TokenKind.Gte,
  "&&": TokenKind.And,
  "||": TokenKind.Or,
  "++": TokenKind.PlusPlus,
  "--": TokenKind.MinusMinus,
  "+=": TokenKind.PlusEq,
  "-=": TokenKind.MinusEq,
  "*=": TokenKind.StarEq,
  "/=": TokenKind.SlashEq,
  "<-": TokenKind.Arrow,
  "::": TokenKind.DoubleColon,
};

const ONE_CHAR_TOKENS: Record<string, TokenKind> = {
  "+": TokenKind.Plus,
  "-": TokenKind.Minus,
  "*": TokenKind.Star,
  "/": TokenKind.Slash,
  "%": TokenKind.Percent,
  "!": TokenKind.Not,
  "&": TokenKind.Ampersand,
  "=": TokenKind.Assign,
  "<": TokenKind.Lt,
  ">": TokenKind.Gt,
  "(": TokenKind.LParen,
  ")": TokenKind.RParen,
  "{": TokenKind.LBrace,
  "}": TokenKind.RBrace,
  "[": TokenKind.LBracket,
  "]": TokenKind.RBracket,
  ";": TokenKind.Semicolon,
  ",": TokenKind.Comma,
  ":": TokenKind.Colon,
  ".": TokenKind.Dot,
};

export interface LexResult {
  tokens: Token[];
  allTokens: Token[];
  diagnostics: VelaDiagnostic[];
}

export class Lexer {
  private pos = 0;
  private line = 1;
  private column = 1;
  private readonly tokens: Token[] = [];
  private readonly allTokens: Token[] = [];
  private readonly diagnostics: VelaDiagnostic[] = [];

  constructor(
    private readonly text: string,
    private readonly uri: string,
  ) {}

  tokenize(): LexResult {
    while (!this.atEnd()) {
      const consumedTrivia = this.skipWhitespaceAndComments();
      if (consumedTrivia && this.atEnd()) {
        break;
      }
      if (this.atEnd()) {
        break;
      }
      this.scanToken();
    }
    const eofPos = this.currentPosition();
    const eof = this.makeToken(TokenKind.Eof, "", "", eofPos, eofPos);
    this.tokens.push(eof);
    this.allTokens.push(eof);
    return {
      tokens: this.tokens,
      allTokens: this.allTokens,
      diagnostics: this.diagnostics,
    };
  }

  private atEnd(): boolean {
    return this.pos >= this.text.length;
  }

  private peek(offset = 0): string {
    return this.text[this.pos + offset] ?? "\0";
  }

  private currentPosition(): VelaPosition {
    return makePosition(this.line, this.column, this.pos);
  }

  private advance(): string {
    const ch = this.text[this.pos] ?? "\0";
    this.pos += 1;
    if (ch === "\n") {
      this.line += 1;
      this.column = 1;
    } else {
      this.column += 1;
    }
    return ch;
  }

  private makeToken(kind: TokenKind, value: string, lexeme: string, start: VelaPosition, end?: VelaPosition): Token {
    return {
      kind,
      value,
      lexeme,
      range: makeRange(this.uri, start, end ?? this.currentPosition()),
    };
  }

  private pushToken(kind: TokenKind, value: string, lexeme: string, start: VelaPosition): void {
    const token = this.makeToken(kind, value, lexeme, start);
    this.tokens.push(token);
    this.allTokens.push(token);
  }

  private pushTrivia(kind: TokenKind, value: string, lexeme: string, start: VelaPosition): void {
    this.allTokens.push(this.makeToken(kind, value, lexeme, start));
  }

  private addDiagnostic(code: string, message: string, start: VelaPosition, hint?: string): void {
    this.diagnostics.push({
      code,
      message,
      hint,
      severity: DiagnosticSeverity.Error,
      range: makeRange(this.uri, start, this.currentPosition()),
      source: "vela-lsp",
    });
  }

  private skipWhitespaceAndComments(): boolean {
    let consumed = false;
    while (!this.atEnd()) {
      const ch = this.peek();
      if (ch === " " || ch === "\t" || ch === "\r" || ch === "\n") {
        consumed = true;
        this.advance();
        continue;
      }
      if (ch === "/" && this.peek(1) === "/") {
        consumed = true;
        const start = this.currentPosition();
        let lexeme = "";
        while (!this.atEnd() && this.peek() !== "\n") {
          lexeme += this.advance();
        }
        this.pushTrivia(TokenKind.Comment, lexeme, lexeme, start);
        continue;
      }
      if (ch === "#") {
        consumed = true;
        const start = this.currentPosition();
        let lexeme = "";
        while (!this.atEnd() && this.peek() !== "\n") {
          lexeme += this.advance();
        }
        this.pushTrivia(TokenKind.Comment, lexeme, lexeme, start);
        continue;
      }
      break;
    }
    return consumed;
  }

  private scanToken(): void {
    const start = this.currentPosition();
    const two = `${this.peek()}${this.peek(1)}`;
    const twoKind = TWO_CHAR_TOKENS[two];
    if (twoKind) {
      this.advance();
      this.advance();
      this.pushToken(twoKind, two, two, start);
      return;
    }

    const one = this.peek();
    const oneKind = ONE_CHAR_TOKENS[one];
    if (oneKind) {
      this.advance();
      this.pushToken(oneKind, one, one, start);
      return;
    }

    if (one === '"') {
      this.scanString(start);
      return;
    }
    if (one === "'") {
      this.scanChar(start);
      return;
    }
    if (isDigit(one)) {
      this.scanNumber(start);
      return;
    }
    if (isIdentifierStart(one)) {
      this.scanIdentifier(start);
      return;
    }

    const unexpected = this.advance();
    this.addDiagnostic(
      "vela.lex.unexpectedCharacter",
      `unexpected character ${JSON.stringify(unexpected)}`,
      start,
      "remove the character or replace it with a valid Vela token",
    );
  }

  private scanString(start: VelaPosition): void {
    let lexeme = this.advance();
    let value = "";
    while (!this.atEnd() && this.peek() !== '"') {
      if (this.peek() === "\n") {
        this.addDiagnostic(
          "vela.lex.unterminatedString",
          "unterminated string literal",
          start,
          "close the string literal before the end of the line",
        );
        this.pushToken(TokenKind.StringLiteral, value, lexeme, start);
        return;
      }
      if (this.peek() === "\\") {
        lexeme += this.advance();
        if (this.atEnd()) {
          this.addDiagnostic(
            "vela.lex.unterminatedString",
            "unterminated string literal",
            start,
            "complete the escape sequence and close the string literal",
          );
          this.pushToken(TokenKind.StringLiteral, value, lexeme, start);
          return;
        }
        const esc = this.advance();
        lexeme += esc;
        value += decodeEscape(esc, '"');
      } else {
        const ch = this.advance();
        lexeme += ch;
        value += ch;
      }
    }
    if (this.atEnd()) {
      this.addDiagnostic(
        "vela.lex.unterminatedString",
        "unterminated string literal",
        start,
        "close the string literal with a double quote",
      );
      this.pushToken(TokenKind.StringLiteral, value, lexeme, start);
      return;
    }
    lexeme += this.advance();
    this.pushToken(TokenKind.StringLiteral, value, lexeme, start);
  }

  private scanChar(start: VelaPosition): void {
    let lexeme = this.advance();
    let value = "";
    if (this.atEnd() || this.peek() === "'") {
      this.addDiagnostic(
        "vela.lex.emptyChar",
        "empty char literal",
        start,
        "provide exactly one character or escape sequence",
      );
      if (!this.atEnd() && this.peek() === "'") {
        lexeme += this.advance();
      }
      this.pushToken(TokenKind.CharLiteral, value, lexeme, start);
      return;
    }
    if (this.peek() === "\\") {
      lexeme += this.advance();
      if (this.atEnd()) {
        this.addDiagnostic(
          "vela.lex.unterminatedChar",
          "unterminated char literal",
          start,
          "complete the escape sequence and close the character literal",
        );
        this.pushToken(TokenKind.CharLiteral, value, lexeme, start);
        return;
      }
      const esc = this.advance();
      lexeme += esc;
      value = decodeEscape(esc, "'");
    } else {
      if (this.peek() === "\n") {
        this.addDiagnostic(
          "vela.lex.unterminatedChar",
          "unterminated char literal",
          start,
          "close the character literal before the end of the line",
        );
        this.pushToken(TokenKind.CharLiteral, value, lexeme, start);
        return;
      }
      const ch = this.advance();
      lexeme += ch;
      value = ch;
    }
    if (this.atEnd() || this.peek() !== "'") {
      this.addDiagnostic(
        "vela.lex.unterminatedChar",
        "unterminated char literal",
        start,
        "close the character literal with a single quote",
      );
      this.pushToken(TokenKind.CharLiteral, value, lexeme, start);
      return;
    }
    lexeme += this.advance();
    if (value.codePointAt(0) !== undefined && value.codePointAt(0)! > 0xff) {
      this.addDiagnostic(
        "vela.lex.charOutOfRange",
        `char literal U+${value.codePointAt(0)!.toString(16).toUpperCase().padStart(4, "0")} is outside the U8 range`,
        start,
        "Char is U8; use a character with codepoint 0..255",
      );
    }
    this.pushToken(TokenKind.CharLiteral, value, lexeme, start);
  }

  private scanNumber(start: VelaPosition): void {
    const begin = this.pos;
    if (this.peek() === "0" && (this.peek(1) === "x" || this.peek(1) === "X")) {
      this.advance();
      this.advance();
      while (!this.atEnd() && /[0-9a-fA-F_]/.test(this.peek())) {
        this.advance();
      }
      const lexeme = this.text.slice(begin, this.pos);
      this.appendNumber(TokenKind.IntLiteral, lexeme, start, validHexInteger(lexeme));
      return;
    }
    if (this.peek() === "0" && (this.peek(1) === "b" || this.peek(1) === "B")) {
      this.advance();
      this.advance();
      while (!this.atEnd() && /[01_]/.test(this.peek())) {
        this.advance();
      }
      const lexeme = this.text.slice(begin, this.pos);
      this.appendNumber(TokenKind.IntLiteral, lexeme, start, validBinaryInteger(lexeme));
      return;
    }

    while (!this.atEnd() && /[0-9_]/.test(this.peek())) {
      this.advance();
    }
    let isFloatLiteral = false;
    if (!this.atEnd() && this.peek() === "." && this.peek(1) !== ".") {
      isFloatLiteral = true;
      this.advance();
      while (!this.atEnd() && /[0-9_]/.test(this.peek())) {
        this.advance();
      }
    }
    if (!this.atEnd() && (this.peek() === "e" || this.peek() === "E")) {
      isFloatLiteral = true;
      this.advance();
      if (!this.atEnd() && (this.peek() === "+" || this.peek() === "-")) {
        this.advance();
      }
      while (!this.atEnd() && /[0-9_]/.test(this.peek())) {
        this.advance();
      }
    }

    const lexeme = this.text.slice(begin, this.pos);
    if (isFloatLiteral) {
      this.appendNumber(TokenKind.FloatLiteral, lexeme, start, validFloat(lexeme));
    } else {
      this.appendNumber(TokenKind.IntLiteral, lexeme, start, validDecimalInteger(lexeme));
    }
  }

  private appendNumber(kind: TokenKind.IntLiteral | TokenKind.FloatLiteral, lexeme: string, start: VelaPosition, valid: boolean): void {
    if (!valid) {
      this.addDiagnostic(
        "vela.lex.invalidNumber",
        `invalid numeric literal ${JSON.stringify(lexeme)}`,
        start,
        "check the digits, prefix, exponent, and underscore separators",
      );
    }
    const normalized = lexeme.replaceAll("_", "");
    if (kind === TokenKind.IntLiteral && valid) {
      const value = parseInteger(normalized);
      if (value > 0xffff) {
        this.addDiagnostic(
          "vela.lex.integerOutOfRange",
          `integer literal ${value} is outside the 16-bit range`,
          start,
          "use a value between 0 and 65535",
        );
      }
    }
    this.pushToken(kind, lexeme, lexeme, start);
  }

  private scanIdentifier(start: VelaPosition): void {
    const begin = this.pos;
    while (!this.atEnd() && isIdentifierContinue(this.peek())) {
      this.advance();
    }
    const text = this.text.slice(begin, this.pos);
    this.pushToken(KEYWORDS[text] ?? TokenKind.Identifier, text, text, start);
  }
}

function isDigit(ch: string): boolean {
  return ch >= "0" && ch <= "9";
}

function isIdentifierStart(ch: string): boolean {
  return /[A-Za-z_]/.test(ch);
}

function isIdentifierContinue(ch: string): boolean {
  return /[A-Za-z0-9_]/.test(ch);
}

function decodeEscape(ch: string, quote: '"' | "'"): string {
  const escapes: Record<string, string> = {
    n: "\n",
    t: "\t",
    "\\": "\\",
    "0": "\0",
  };
  escapes[quote] = quote;
  return escapes[ch] ?? ch;
}

function validUnderscoreDigits(text: string, digitPattern: string): boolean {
  return new RegExp(`^[${digitPattern}]+(?:_[${digitPattern}]+)*$`).test(text);
}

function validDecimalInteger(text: string): boolean {
  return validUnderscoreDigits(text, "0-9");
}

function validHexInteger(text: string): boolean {
  return /^0[xX]/.test(text) && validUnderscoreDigits(text.slice(2), "0-9a-fA-F");
}

function validBinaryInteger(text: string): boolean {
  return /^0[bB]/.test(text) && validUnderscoreDigits(text.slice(2), "01");
}

function validFloat(text: string): boolean {
  const dec = String.raw`\d+(?:_\d+)*`;
  const body = new RegExp(String.raw`^(?:${dec}\.${dec}?|${dec}?\.${dec})(?:[eE][+-]?${dec})?$`);
  const expOnly = new RegExp(String.raw`^${dec}[eE][+-]?${dec}$`);
  return (body.test(text) || expOnly.test(text)) && Number.isFinite(Number(text.replaceAll("_", "")));
}

function parseInteger(normalized: string): number {
  if (normalized.startsWith("0x") || normalized.startsWith("0X")) {
    return Number.parseInt(normalized.slice(2), 16);
  }
  if (normalized.startsWith("0b") || normalized.startsWith("0B")) {
    return Number.parseInt(normalized.slice(2), 2);
  }
  return Number.parseInt(normalized, 10);
}
